hc_dpki/src/seed.rs

```diff
@@ -0,0 +1,558 @@
+    pub fn get_bundle(
+        &mut self,
+        passphrase: String,
+        hint: String,
+        config: Option<PwHashConfig>,
+    ) -> Result<bundle::KeyBundle, HolochainError> {
+        Ok(self.s.get_seed_bundle(passphrase, hint, config)?)
+    }
+
+    /// delegate to base struct
+    pub fn new(s: SecBuf) -> Self {
+        DevicePinSeed {
+            s: Seed::new(&"hcDevicePinSeed".to_string(), SeedInit(s)),
+        }
+    }
+
+    /// generate an application keypair given an index based on this seed
+    /// @param {number} index
+    /// @return {Keypair}
+    pub fn get_application_keypair(&mut self, index: u64) -> Result<Keypair, HolochainError> {
+        (index >= 1).ok_or(HolochainError::ErrorGeneric("Invalid index".to_string()))?;
+
+        let mut out_seed = SecBuf::with_insecure(32);
+        let mut placeholder = SecBuf::with_insecure_from_string("HCAPPLIC".to_string());
+        kdf::derive(
+            &mut out_seed,
+            index.clone(),
+            &mut placeholder,
+            &mut self.s.seed_buf,
+        )?;
+
+        Ok(Keypair::new_from_seed(&mut out_seed)?)
+    }
+}
+#[derive(Debug)]
+pub struct DeviceSeed {
+    s: Seed,
+}
+
+impl DeviceSeed {
+    /// creates bundle for for the seed
+    pub fn get_bundle(
+        &mut self,
+        passphrase: String,
+        hint: String,
+        config: Option<PwHashConfig>,
+    ) -> Result<bundle::KeyBundle, HolochainError> {
+        Ok(self.s.get_seed_bundle(passphrase, hint, config)?)
+    }
+
+    /// delegate to base struct
+    pub fn new(s: SecBuf) -> Self {
+        DeviceSeed {
+            s: Seed::new(&"hcDeviceSeed".to_string(), SeedInit(s)),
+        }
+    }
+
+    /// generate a device pin seed by applying pwhash of pin with this seed as the salt
+    /// @param {string} pin - should be >= 4 characters 1-9
+    /// @return {DevicePinSeed}
+    pub fn get_device_pin_seed(
+        &mut self,
+        pin: String,
+        config: Option<PwHashConfig>,
+    ) -> Result<DevicePinSeed, HolochainError> {
+        (pin.len() >= 4).ok_or(HolochainError::ErrorGeneric("Invalid PIN Size".to_string()))?;
+
+        // let pin_encoded = base64::encode(&pin);
+        let mut pin_buf = SecBuf::with_insecure_from_string(pin);
+
+        let mut hash = SecBuf::with_insecure(pwhash::HASHBYTES);
+
+        util::pw_hash(&mut pin_buf, &mut self.s.seed_buf, &mut hash, config)?;
+
+        Ok(DevicePinSeed::new(hash))
+    }
+}
+
+#[derive(Debug)]
+pub struct RootSeed {
+    s: Seed,
+}
+
+impl RootSeed {
+    /// creates bundle for for the seed
+    pub fn get_bundle(
+        &mut self,
+        passphrase: String,
+        hint: String,
+        config: Option<PwHashConfig>,
+    ) -> Result<bundle::KeyBundle, HolochainError> {
+        Ok(self.s.get_seed_bundle(passphrase, hint, config)?)
+    }
+
+    /// Get a new, completely random root seed
+    pub fn new_random() -> Self {
+        let mut s = SecBuf::with_insecure(32);
```

Comment:
I vote for deleting the `new_random()` function. We want the implementor to be able to decide if it should be a secure or insecure buffer, which is provided by the `new` function below. They can randomize that buffer before passing it in.

Reply:
I agree. It makes sense

Reply:
Done :+1: